File: problems/0010_regular-expression-matching/bottom-to-top/Solution.rs
impl Solution {
    pub fn is_match(s: String, p: String) -> bool {
        let (s_len, p_len, s_ptr, p_ptr) = (s.len(), p.len(), s.as_ptr(), p.as_ptr());
        let mut memo = vec![vec![false; p_len + 1]; s_len + 1];
        memo[s_len][p_len] = true;
        for i in (0..(s_len + 1)).rev() {
            for j in (0..p_len).rev() {
                let match_char = (i < s_len &&
                        (unsafe { *(p_ptr.add(j)) == *(s_ptr.add(i)) } ||
                            unsafe { *(p_ptr.add(j)) } as char == '.'));
                memo[i][j] = if j + 1 < p_len && unsafe { *(p_ptr.add(j + 1)) } as char == '*' {
                            (memo[i][j + 2] || (match_char && memo[i + 1][j]))
                        } else {
                            (match_char && memo[i + 1][j + 1])
                        };
            }
        }
        memo[0][0]
    }
}

File: problems/0010_regular-expression-matching/bottom-to-top/solution.ts
function isMatch(s: string, p: string): boolean {
    const memo = new Array<Array<boolean>>(s.length + 1)
    for (let i = 0; i < memo.length; ++i)
        memo[i] = new Array<boolean>(p.length + 1)
    memo[s.length][p.length] = true
    for (let i = s.length; i >= 0; --i)
        for (let j = p.length - 1; j >= 0; --j) {
            const matchChar = (i < s.length &&
                    (p[j] === s[i] || p[j] === '.'))
            memo[i][j] = (j < p.length - 1 && p[j + 1] === '*' ?
                    (memo[i][j + 2] || (matchChar && memo[i + 1][j])) :
                    (matchChar && memo[i + 1][j + 1]))
        }
    return Boolean(memo[0][0])
};
